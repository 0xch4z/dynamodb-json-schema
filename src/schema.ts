import { JSONSchema4 } from 'json-schema';

export const ERR_SCHEMA_MUST_BE_OBJECT = `The JSON Schema for a table item must be of type 'object'`;
export const ERR_SCHEMA_MUST_HAVE_PROPERTIES = `The JSON Schema for a table item must have properties`;

export type JSONSchema = JSONSchema4 & {
  dynamoDBScalarType?: string;
};

export type JSONSchemaObject = JSONSchema & {
  type: 'object';
  properties: Record<string, JSONSchema>;
};

export function assertJSONSchemaObject(schema: JSONSchema) {
  if (schema.type !== 'object') {
    throw new Error(ERR_SCHEMA_MUST_BE_OBJECT);
  }

  if (!schema.properties || !Object.keys(schema.properties).length) {
    throw new Error(ERR_SCHEMA_MUST_HAVE_PROPERTIES);
  }

  return schema as JSONSchemaObject;
}
