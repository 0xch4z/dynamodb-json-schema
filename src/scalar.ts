import { JSONSchema } from './schema';

export const enum DynamoDBIndexableScalarType {
  Number = 'N',
  String = 'S',
  Binary = 'B'
}

export function mustIndexableScalarType(identifier: string) {
  switch (identifier) {
    case DynamoDBIndexableScalarType.Binary:
      return DynamoDBIndexableScalarType.Binary;
    case DynamoDBIndexableScalarType.Number:
      return DynamoDBIndexableScalarType.Number;
    case DynamoDBIndexableScalarType.String:
      return DynamoDBIndexableScalarType.String;
    default:
      throw new Error(`Invalid DynamoDB indexable scalar type ${identifier}`);
  }
}

export function mustInferIndexableScalarType({
  dynamoDBScalarType,
  type
}: JSONSchema): DynamoDBIndexableScalarType {
  if (dynamoDBScalarType) {
    return mustIndexableScalarType(dynamoDBScalarType);
  }

  switch (type) {
    case 'integer':
    case 'number':
      return DynamoDBIndexableScalarType.Number;
    case 'string':
      return DynamoDBIndexableScalarType.String;
    default:
      throw new Error(
        `Cannot infer indexable scalar type from JSON Schema type '${type}'`
      );
  }
}
