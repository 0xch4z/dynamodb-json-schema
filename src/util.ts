export function pluckUndefinedObjectValues<T extends {}>(subject: T) {
  Object.keys(subject).forEach(
    key => subject[key] === undefined && delete subject[key]
  );
}
