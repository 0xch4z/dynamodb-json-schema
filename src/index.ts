import {
  AttributeDefinitions,
  KeySchema,
  GlobalSecondaryIndex,
  LocalSecondaryIndex,
  Projection
} from "aws-sdk/clients/dynamodb";
import { JSONSchema, assertJSONSchemaObject, JSONSchemaObject } from "./schema";

import { mustInferIndexableScalarType } from "./scalar";
import { pluckUndefinedObjectValues } from "./util";

export interface TableSchema {
  AttributeDefintions: AttributeDefinitions;
  TableName?: string;
  KeySchema: KeySchema;
  LocalSecondaryIndexes?: LocalSecondaryIndex[];
  GlobalSecondaryIndexes?: GlobalSecondaryIndex[];
}

export interface IndexDefinition {
  hashKey?: string;
  rangeKey?: string;
}

export interface SecondaryIndexDefinition {
  indexName: string;
  projectionType?: string;
  nonKeyAttributes?: string[];
}

export interface GlobalSecondaryIndexDefiniton
  extends SecondaryIndexDefinition {
  hashKey: string;
  rangeKey?: string;
}

export interface LocalSecondaryIndexDefinition
  extends SecondaryIndexDefinition {
  rangeKey: string;
}

export interface SecondaryIndex {
  IndexName: string;
  KeySchema: KeySchema;
  Projection: Projection;
}

export interface GetTableSchemaOptions {
  tableName?: string;
  hashKey: string;
  rangeKey?: string;
  gsis?: GlobalSecondaryIndexDefiniton[];
  lsis?: LocalSecondaryIndexDefinition[];
  itemSchema: JSONSchema;
}

export const enum IndexKeyType {
  Hash = "HASH",
  Range = "RANGE"
}

export function getUniqueIndexedAttributes(indexes: IndexDefinition[]) {
  return indexes
    .reduce<Array<string | undefined>>(
      (acc, curr) => [...acc, curr.hashKey, curr.rangeKey],
      []
    )
    .filter(Boolean)
    .filter((name, i, a) => a.indexOf(name) === i);
}

function getKeySchemaFromIndexDefinition(index: IndexDefinition): KeySchema {
  return [
    index.hashKey && {
      AttributeName: index.hashKey,
      KeyType: IndexKeyType.Hash
    },
    index.rangeKey && {
      AttributeName: index.rangeKey,
      KeyType: IndexKeyType.Range
    }
  ].filter(Boolean);
}

function getAttributeDefinitionsFromSchema(
  indexedAttributes: string[],
  schema: JSONSchemaObject
) {
  return indexedAttributes.map(attribute => {
    const attributeSchema = schema.properties[attribute];
    if (typeof attributeSchema !== "object") {
      throw new Error(
        `Could not find schema definiton for specified indexed attribute '${attribute}'`
      );
    }

    const type = mustInferIndexableScalarType(attributeSchema);
    return { AttributeName: attribute, AttributeType: type };
  });
}

function getSecondaryIndexFromDefinition<
  T extends SecondaryIndexDefinition & IndexDefinition
>(definition: T): SecondaryIndex {
  return {
    IndexName: definition.indexName,
    KeySchema: getKeySchemaFromIndexDefinition(definition),
    Projection: {
      ProjectionType: definition.projectionType,
      NonKeyAttributes: definition.nonKeyAttributes
    }
  };
}

export function getTableSchema({
  tableName,
  itemSchema,
  hashKey,
  rangeKey,
  gsis = [],
  lsis = []
}: GetTableSchemaOptions): TableSchema {
  const schemaObject = assertJSONSchemaObject(itemSchema);
  const indexes = [{ hashKey, rangeKey }, ...gsis, ...lsis];
  const indexedAttributes = getUniqueIndexedAttributes(indexes);
  const keySchema = getKeySchemaFromIndexDefinition({ hashKey, rangeKey });
  const attributes = getAttributeDefinitionsFromSchema(
    indexedAttributes,
    schemaObject
  );

  const tableSchema: TableSchema = {
    TableName: tableName,
    AttributeDefintions: attributes,
    KeySchema: keySchema
  };

  if (gsis.length) {
    tableSchema.GlobalSecondaryIndexes = gsis.map(
      getSecondaryIndexFromDefinition
    );
  }

  if (lsis.length) {
    tableSchema.LocalSecondaryIndexes = lsis.map(
      getSecondaryIndexFromDefinition
    );
  }

  pluckUndefinedObjectValues(tableSchema);
  return tableSchema;
}
