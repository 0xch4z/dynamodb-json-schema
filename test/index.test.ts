import { getTableSchema } from '../src';
import { ERR_SCHEMA_MUST_BE_OBJECT, JSONSchema } from '../src/schema';

describe('getTableSchema', () => {
  test('throws error on invalid JSON Schema object', () => {
    expect(() =>
      getTableSchema({
        tableName: 'some-table',
        hashKey: 'someHashKey',
        itemSchema: { type: 'boolean' } // should be an object
      })
    ).toThrowError(ERR_SCHEMA_MUST_BE_OBJECT);
  });

  test('throws error when indexed attribute not included in JSON schema', () => {
    expect(() =>
      getTableSchema({
        tableName: 'some-table',
        hashKey: 'someKeyNotInSchema',
        itemSchema: {
          type: 'object',
          properties: {
            someIndexedAttribute: { type: 'string' }
          }
        }
      })
    ).toThrowError(
      `Could not find schema definiton for specified indexed attribute 'someKeyNotInSchema'`
    );
  });

  test('should return proper table schema for valid JSON schema', () => {
    const itemSchema: JSONSchema = {
      type: 'object',
      properties: {
        myHashKey: { type: 'string' }
      }
    };
    const tableSchema = getTableSchema({
      tableName: 'some-table',
      hashKey: 'myHashKey',
      itemSchema
    });

    expect(tableSchema).toMatchSnapshot();
  });

  test('should return proper table schema for valid JSON schema with secondary indexes', () => {
    const itemSchema: JSONSchema = {
      type: 'object',
      properties: {
        id: { type: 'string' },
        login: { type: 'string' },
        name: { type: 'string' },
        age: { type: 'integer' }
      }
    };

    const tableSchema = getTableSchema({
      tableName: 'user-table',
      hashKey: 'id',
      gsis: [
        {
          indexName: 'byLogin',
          hashKey: 'login',
          projectionType: 'ALL'
        }
      ],
      lsis: [
        {
          indexName: 'byName',
          rangeKey: 'name',
          projectionType: 'ALL'
        }
      ],
      itemSchema
    });

    expect(tableSchema).toMatchSnapshot();
  });
});
