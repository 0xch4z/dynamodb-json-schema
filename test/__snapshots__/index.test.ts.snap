// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTableSchema should return proper table schema for valid JSON schema 1`] = `
Object {
  "AttributeDefintions": Array [
    Object {
      "AttributeName": "myHashKey",
      "AttributeType": "S",
    },
  ],
  "KeySchema": Array [
    Object {
      "AttributeName": "myHashKey",
      "KeyType": "HASH",
    },
  ],
  "TableName": "some-table",
}
`;

exports[`getTableSchema should return proper table schema for valid JSON schema with secondary indexes 1`] = `
Object {
  "AttributeDefintions": Array [
    Object {
      "AttributeName": "id",
      "AttributeType": "S",
    },
    Object {
      "AttributeName": "login",
      "AttributeType": "S",
    },
    Object {
      "AttributeName": "name",
      "AttributeType": "S",
    },
  ],
  "GlobalSecondaryIndexes": Array [
    Object {
      "IndexName": "byLogin",
      "KeySchema": Array [
        Object {
          "AttributeName": "login",
          "KeyType": "HASH",
        },
      ],
      "Projection": Object {
        "NonKeyAttributes": undefined,
        "ProjectionType": "ALL",
      },
    },
  ],
  "KeySchema": Array [
    Object {
      "AttributeName": "id",
      "KeyType": "HASH",
    },
  ],
  "LocalSecondaryIndexes": Array [
    Object {
      "IndexName": "byName",
      "KeySchema": Array [
        Object {
          "AttributeName": "name",
          "KeyType": "RANGE",
        },
      ],
      "Projection": Object {
        "NonKeyAttributes": undefined,
        "ProjectionType": "ALL",
      },
    },
  ],
  "TableName": "user-table",
}
`;
